File: go/stack/stack.go
/*
A stack is a linear data structure that operates on the principle of LIFO (Last In, First Out).
This means that the last element added to the stack is the first one to be removed. You can visualize
a stack like a stack of plates: you add plates to the top, and the only plate you can remove is the one on the top.

The fundamental operations of a stack include:

Push: Add an element to the top of the stack.
Pop: Remove the element from the top of the stack.
Peek (or Top): Retrieve the top element without removing it.
IsEmpty: Check whether the stack is empty.
*/

/*
Stack (LIFO - Last In, First Out)

    |     |
    |     |
    |  5  |  <-- Top
    |  3  |
    |  1  |
    +-----+

    Push(7)
    |     |
    |  7  |  <-- Top
    |  5  |
    |  3  |
    |  1  |
    +-----+

    Pop() -> 7
    |     |
    |     |
    |  5  |  <-- Top
    |  3  |
    |  1  |
    +-----+
*/

package main

import "fmt"

type node struct {
	value int
	next  *node
}

type Stack struct {
	top *node
}

func (s *Stack) Push(value int) {
	newNode := &node{value: value}
	if s.top == nil {
		s.top = newNode
	} else {
		newNode.next = s.top
		s.top = newNode
	}
}

func (s *Stack) Pop() int {
	if s.top == nil {
		return -1
	}

	value := s.top.value

	s.top = s.top.next

	return value
}

func (s *Stack) Peek() int {
	if s.top == nil {
		return -1
	}

	return s.top.value
}

func (s *Stack) IsEmpty() bool {
	return s.top == nil
}

func (s *Stack) PrintStack() {
	current := s.top
	for current != nil {
		fmt.Println(current.value)
		current = current.next
	}
}

func main() {

	stack := &Stack{}

	stack.Push(10)
	stack.Push(8)
	stack.Push(9)

	stack.PrintStack()
}

File: c++/stack/stack.cpp
#include <iostream>
class Node
{
public:
    int value;
    Node *next;
    Node(int value) : value(value), next(nullptr) {}
};

class Stack
{
private:
    Node *top;
    int height;

public:
    Stack() : top(nullptr), height(0) {}

    void printStack()
    {
        Node *current = top;
        while (current)
        {
            std::cout << current->value << std::endl;
            current = current->next;
        }
    }

    void push(int value)
    {
        Node *newNode = new Node(value);
        newNode->next = top;
        top = newNode;
        height++;
    }

    void pop()
    {
        if (top == nullptr)
            return;
        Node *newNode = top;
        top = top->next;
        delete newNode;
        height--;
    }

    bool isEmpty()
    {
        return height == 0;
    }
    int getTop()
    {
        if (top == nullptr)
            return -1;
        return top->value;
    }

    int getSize()
    {
        return height;
    }
};